app/controllers/api/grade_entry_forms_controller.rb

```diff
@@ -31,6 +31,109 @@ def index
+    #     :name, :out_of, :bonus
+    def create
+      if has_missing_params?([:short_identifier])
+        # incomplete/invalid HTTP params
+        render 'shared/http_status', locals: { code: '422', message:
+          HttpStatusHelper::ERROR_CODE['message']['422'] }, status: 422
+        return
+      end
+
+      # check if there is an existing assignment
+      form = GradeEntryForm.find_by_short_identifier(params[:short_identifier])
+      unless form.nil?
+        render 'shared/http_status', locals: { code: '409', message:
+          'Grade Entry Form already exists' }, status: 409
+        return
+      end
+
+      ApplicationRecord.transaction do
+        create_params = params.permit(*DEFAULT_FIELDS, :show_total)
+        create_params[:is_hidden] ||= false
+        create_params[:description] ||= ''
+        new_form = GradeEntryForm.new(create_params)
+        unless new_form.save
+          render 'shared/http_status', locals: { code: '500', message:
+            new_form.errors.full_messages.first }, status: 500
+          raise ActiveRecord::Rollback
+        end
+
+        params[:grade_entry_items]&.each&.with_index do |column_params, i|
+          column_params = column_params.permit(:name, :out_of, :bonus).to_h.symbolize_keys
+          grade_item = GradeEntryItem.new(**column_params, grade_entry_form_id: new_form.id, position: i + 1)
+          unless grade_item.save
+            render 'shared/http_status', locals: { code: '500', message:
+              grade_item.errors.full_messages.first }, status: 500
+            raise ActiveRecord::Rollback
+          end
+        end
+        render 'shared/http_status', locals: { code: '200', message:
+          HttpStatusHelper::ERROR_CODE['message']['200'] }, status: 200
+      end
+    end
+
+    # create a new grade entry form
+    # params:
+    #   :short_identifier, :description, :date, :is_hidden
+    #   grade_entry_items:
+    #     :id, :name, :out_of, :bonus
+    #
+    # if the grade_entry_items id param is set, an existing item will be
+    # updated, otherwise a new grade_entry_item will be created
+    def update
+      # check if there is an existing assignment
+      form = GradeEntryForm.find_by_id(params[:id])
+      if form.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          'Grade Entry Form not found' }, status: 404
+        return
+      end
+
+      ApplicationRecord.transaction do
+        update_params = params.permit(*DEFAULT_FIELDS, :show_total)
+        unless form.update(update_params)
+          render 'shared/http_status', locals: { code: '500', message:
+            form.errors.full_messages.first }, status: 500
+          raise ActiveRecord::Rollback
+        end
+
+        params[:grade_entry_items]&.each do |column_params|
+          if column_params[:id].nil?
+            column_params = column_params.permit(:name, :out_of, :bonus).to_h.symbolize_keys
+            position = form.grade_entry_items.count + 1
```

Comment:
Initialize this once based on the count and then increment it manually in the loop (saves the extra DB queries).